ev3dev2/stopwatch.py

```diff
@@ -9,76 +9,113 @@
 else:
     import datetime as dt
 
-
 def get_ticks_ms():
     if is_micropython():
         return utime.ticks_ms()
     else:
         return int(dt.datetime.timestamp(dt.datetime.now()) * 1000)
 
-
 class StopWatch(object):
-
+    """
+    A timer class which lets you start timing and then check the amount of time
+    elapsed.
+    """
     def __init__(self, desc=None):
+        """
+        Initializes the StopWatch but does not start it.
+
+        desc:
+            A string description to print when stringifying.
+        """
         self.desc = desc
-        self._value = 0
-        self.start_time = None
-        self.prev_update_time = None
+        self._start_time = None
+        self._stopped_total_time = None
 
     def __str__(self):
-        if self.desc is not None:
-            return self.desc
-        else:
-            return self.__class__.__name__
+        name = self.desc if self.desc is not None else self.__class__.__name__
+        return "{}: {}".format(name, self.hms_str)
 
     def start(self):
-        assert self.start_time is None, "%s is already running" % self
-        self.start_time = get_ticks_ms()
-
-    def update(self):
-
-        if self.start_time is None:
-            return
-
-        current_time = get_ticks_ms()
-
-        if self.prev_update_time is None:
-            delta = current_time - self.start_time
-        else:
-            delta = current_time - self.prev_update_time
-
-        self._value += delta
-        self.prev_update_time = current_time
+        """
+        Starts the timer. If the timer is already running, resets it.
+        """
+        self._stopped_total_time = None
+        self._start_time = get_ticks_ms()
 
     def stop(self):
-
-        if self.start_time is None:
+        """
+        Stops the timer. The time value of this Stopwatch is paused and will not continue increasing.
+        """
+        if self._start_time is None:
             return
 
-        self.update()
-        self.start_time = None
-        self.prev_update_time = None
+        self._stopped_total_time = get_ticks_ms() - self._start_time
+        self._start_time = None
 
     def reset(self):
-        self.stop()
-        self._value = 0
+        """
+        Resets the timer, and starts it again.
+        """
+        self._start_time = None
+        self._stopped_total_time = None
+        self.start()
```

Comment:
Having the reset call start the time feels odd. How about we drop the start() here but then add a restart method that calls reset() followed by start()?

Reply:
"restart" does make sense to me. I think that's a better option than what's currently here.